### PR TITLE
Don't notify non-board members of card changes

### PR DESCRIPTION
**Pull request by wiggin77**
#### Summary
Don't notify non-board members of card changes 

- include permissions service in notifications backends 
- use permissions service to ensure @mentions are to users on team 
- use permissions service to ensure subscribers are members of board

**Note:  PM decision pending on whether mentioned users get automatically added to board when they get subscribed to cards**

#### Ticket Link
fixes https://github.com/mattermost/focalboard/issues/2666

### CODE REVIEW

File: server/services/notify/plugindelivery/plugin_delivery.go [hunk 1]
```diff
@@ -42,45 +39,17 @@ type PluginAPI interface {
...
-func New(botID string, serverRoot string, api PluginAPI) *PluginDelivery {
+func New(botID string, serverRoot string, api PluginAPI, permissions permissions.PermissionsService) *PluginDelivery {
 	return &PluginDelivery{
-		botID:      botID,
-		serverRoot: serverRoot,
-		api:        api,
-	}
-}
-
-func (pd *PluginDelivery) Deliver(mentionUsername string, extract string, evt notify.BlockChangeEvent) error {
```
Comment (wiggin77) [inline server/services/notify/plugindelivery/plugin_delivery.go:58]: This method had been moved/copied during a previous refactor but left here unintentionally. It is not called.

File: server/services/notify/notifysubscriptions/subscriptions_backend.go [hunk 2]
```diff
@@ -179,6 +183,10 @@ func (b *Backend) OnMention(userID string, evt notify.BlockChangeEvent) {
 		return
 	}
 
+	// TODO: Automatically add user to board?  Fail and show UI?  Waiting for PM decision.
```
Comment (jespino) [inline server/services/notify/notifysubscriptions/subscriptions_backend.go:186]: @chenilim @wuwinson we should make a decision here, for consistency with Mattermost Channels, we should ask the user in the frontend for that, if the user has the right permissions to do it.

File: server/services/notify/plugindelivery/mention_deliver.go [hunk 3]
```diff
@@ -40,5 +50,5 @@ func (pd *PluginDelivery) MentionDeliver(mentionUsername string, extract string,
 		ChannelId: channel.Id,
 		Message:   formatMessage(author.Username, extract, evt.Card.Title, link, evt.BlockChanged),
 	}
-	return member.UserId, pd.api.CreatePost(post)
+	return user.Id, pd.api.CreatePost(post)
```
Comment (jespino) [inline server/services/notify/plugindelivery/mention_deliver.go:53]: Should I check if the user has permission to view the board to receive mentions?

File: server/services/notify/plugindelivery/mention_deliver.go [hunk 4]
```diff
@@ -24,12 +29,17 @@ func (pd *PluginDelivery) MentionDeliver(mentionUsername string, extract string,
 		}
 	}
 
+	// make sure mentioned user has permissions to team.
+	if !pd.permissions.HasPermissionToTeam(user.Id, evt.TeamID, model.PermissionViewTeam) {
```
Comment (jespino) [inline server/services/notify/plugindelivery/mention_deliver.go:33]: I'm not sure if this should be the ViewBoard permission, instead of the view team.

**Review by jespino — CHANGES_REQUESTED** [review]
Some doubts around the mentions and permissions checking, otherwise looks good to me.

File: server/services/notify/plugindelivery/mention_deliver.go (hunk 4 above)
Comment (wiggin77) [reply server/services/notify/plugindelivery/mention_deliver.go:33]: That would mean you can only mention users who are already members of the board.  Doesn't seem right to me.

**Comment by wiggin77** [comment]
> Some doubts around the mentions and permissions checking

Thanks @jespino.  The main purpose of this PR was to ensure permissions were correct for receiving change notifications and adding the plumbing for permissions to the notification services.  The spec for mentions wasn't ready when this PR was created so I left a placeholder for that.  With this PR the mention permissions is the same as before:  users can mention anyone on the team.

Another PR will implement the new mentions spec which will address your concerns.

File: server/services/notify/plugindelivery/mention_deliver.go (hunk 4 above)
Comment (jespino) [reply server/services/notify/plugindelivery/mention_deliver.go:33]: Yes, I think you are right.

**Pull request merged**